File: ios/Classes/TCConsentPlugin.m
#import "TCConsentPlugin.h"
#if __has_include(<TCConsent_IAB/TCPrivacyCallbacks.h>)
#import <TCConsent_IAB/TCPrivacyCallbacks.h>
#import <TCConsent_IAB/TCConsentAPI.h>
#import <TCConsent_IAB/TCConsentConstants.h>
#import <TCConsent_IAB/TCPrivacyCenterViewController.h>
#import <TCConsent_IAB/TCMobileConsent.h>
#else
#import <TCConsent/TCPrivacyCallbacks.h>
#import <TCConsent/TCConsentAPI.h>
#import <TCConsent/TCConsentConstants.h>
#import <TCConsent/TCPrivacyCenterViewController.h>
#import <TCConsent/TCMobileConsent.h>
#endif

@interface TCConsentPlugin ()

@property (nonatomic, retain) FlutterMethodChannel* channel;
@property (nonatomic, assign) BOOL blockIOSPrivacyCenterDropOut;

@end

@implementation TCConsentPlugin

+ (void) registerWithRegistrar: (NSObject<FlutterPluginRegistrar>*) registrar
{
    TCConsentPlugin* instance = [[TCConsentPlugin alloc] init];
    instance.channel = [FlutterMethodChannel methodChannelWithName: @"tc_consent_flutter_plugin"
                                                                binaryMessenger: [registrar messenger]];
    [registrar addMethodCallDelegate: instance channel: instance.channel];
}

- (void) handleMethodCall: (FlutterMethodCall*) call result: (FlutterResult) result
{
  if ([@"setSiteIDPrivacyID" isEqualToString:call.method])
  {
      int siteID = [[call.arguments objectForKey: @"siteID"] intValue];
      int privacyID = [[call.arguments objectForKey: @"privacyID"] intValue];
      [[TCMobileConsent sharedInstance] setSiteID: siteID andPrivacyID: privacyID];
      [[TCMobileConsent sharedInstance] registerCallback: self];
      result(@{@"user" : [self getTCUserDictionary]});
  }
  else if ([@"showPrivacyCenter" isEqualToString:call.method])
  {
      TCPrivacyCenterViewController *PCM = [[TCPrivacyCenterViewController alloc] init];
      UIViewController *viewController = ((UIApplication *)[UIApplication sharedApplication]).delegate.window.rootViewController;

      if (@available(iOS 13.0, *))
      {
          [PCM setModalInPresentation: self.blockIOSPrivacyCenterDropOut];
      }
      else
      {
          [[TCLogger sharedInstance] logMessage: @"iOS 13 or higher API not available ! can't block modal Privacy Center." withLevel: TCLogLevel_Error];
      }
      
      [viewController presentViewController: PCM animated: YES completion: nil];
      result(nil);
  }
  else if ([@"acceptAllConsent" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] acceptAllConsent];
      result(nil);
  }
  else if ([@"refuseAllConsent" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] refuseAllConsent];
      result(nil);
  }
  else if ([@"useAcString" isEqualToString:call.method])
  {
      NSNumber *val = [call.arguments objectForKey: @"useAcString"];
      [[TCMobileConsent sharedInstance] useAcString: [val boolValue]];
      result(nil);
  }
  else if ([@"initWithCustomPCM" isEqualToString:call.method])
  {
      int siteID = [[call.arguments objectForKey: @"siteID"] intValue];
      int privacyID = [[call.arguments objectForKey: @"privacyID"] intValue];
      [[TCMobileConsent sharedInstance] customPCMSetSiteID: siteID andPrivacyID: privacyID];
      [[TCMobileConsent sharedInstance] registerCallback: self];
      result(@{@"user" : [self getTCUserDictionary]});
  }
  else if ([@"setConsentDuration" isEqualToString:call.method])
  {
      NSNumber *val = [call.arguments objectForKey: @"months"];
      [[TCMobileConsent sharedInstance] setConsentDuration: [val floatValue]];
      result(nil);
  }
  else if ([@"useCustomPublisherRestrictions" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] useCustomPublisherRestrictions];
      result(nil);
  }
  else if ([@"saveConsentFromPopUp" isEqualToString:call.method])
  {
      NSDictionary *val = [call.arguments objectForKey: @"consent"];
      [[TCMobileConsent sharedInstance] saveConsentFromPopUp: val];
      result(nil);
  }
  else if ([@"saveConsent" isEqualToString:call.method])
  {
      NSDictionary *val = [call.arguments objectForKey: @"consent"];
      [[TCMobileConsent sharedInstance] saveConsent: val];
      result(nil);
  }
  else if ([@"saveConsentFromConsentSourceWithPrivacyAction" isEqualToString:call.method])
  {
      ETCConsentSource source = [[call.arguments objectForKey: @"source"] isEqualToString: @"ETCConsentSource.popUp"] ? Popup : PrivacyCenter;
      NSString *actionValue = [call.arguments objectForKey: @"action"];
      ETCConsentAction action = [actionValue isEqualToString: @"ETCConsentAction.acceptAll"] ? AcceptAll : ([actionValue isEqualToString: @"ETCConsentAction.refuseAll"] ? RefuseAll : Save);
      NSDictionary *consent = [call.arguments objectForKey: @"consent"];
      [[TCMobileConsent sharedInstance] saveConsent: consent fromConsentSource: source withPrivacyAction: action];
      result(@{@"user" : [self getTCUserDictionary]});
  }
  else if ([@"statEnterPCToVendorScreen" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statEnterPCToVendorScreen];
      result(nil);
  }
  else if ([@"statShowVendorScreen" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statShowVendorScreen];
      result(nil);
  }
  else if ([@"statViewPrivacyPoliciesFromPrivacyCenter" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statViewPrivacyPoliciesFromPrivacyCenter];
      result(nil);
  }
  else if ([@"statViewPrivacyCenter" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statViewPrivacyCenter];
      result(nil);
  }
  else if ([@"statViewBanner" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statViewBanner];
      result(nil);
  }
  else if ([@"statViewPrivacyPoliciesFromBanner" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statViewPrivacyPoliciesFromBanner];
      result(nil);
  }
  else if ([@"statViewPrivacyPoliciesFromBanner" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] statViewPrivacyPoliciesFromBanner];
      result(nil);
  }
  else if ([@"getConsentAsJson" isEqualToString:call.method])
  {
      result([[TCMobileConsent sharedInstance] getConsentAsJson]);
  }
  else if ([@"resetSavedConsent" isEqualToString:call.method])
  {
      [[TCMobileConsent sharedInstance] resetSavedConsent];
      result(@{@"user" : [self getTCUserDictionary]});
  }
  else if ([@"setLanguage" isEqualToString:call.method])
  {
      NSString *languageCode = [call.arguments objectForKey: @"languageCode"];
      [[TCMobileConsent sharedInstance] setLanguage: languageCode];
      result(nil);
  }
  else if ([@"blockIOSPrivacyCenterDropOut" isEqualToString:call.method])
  {
      BOOL value = [[call.arguments objectForKey: @"value"] boolValue];
      self.blockIOSPrivacyCenterDropOut = value;
      result(nil);
  }
  else if ([@"isConsentAlreadyGiven" isEqualToString:call.method])
  {
      result([NSNumber numberWithBool: [TCConsentAPI isConsentAlreadyGiven]]);
  }
  else if ([@"getAllAcceptedConsent" isEqualToString:call.method])
  {
      result([TCConsentAPI getAllAcceptedConsent]);
  }
  else if ([@"getLastTimeConsentWasSaved" isEqualToString:call.method])
  {
      result([NSNumber numberWithLong: [TCConsentAPI getLastTimeConsentWasSaved]]);
  }
  else if ([@"isCategoryAccepted" isEqualToString:call.method])
  {
      int ID = [[call.arguments objectForKey: @"id"] intValue];
      result([NSNumber numberWithBool: [TCConsentAPI isCategoryAccepted: ID]]);
  }
  else if ([@"isVendorAccepted" isEqualToString:call.method])
  {
      int ID = [[call.arguments objectForKey: @"id"] intValue];
      result([NSNumber numberWithBool: [TCConsentAPI isVendorAccepted: ID]]);
  }
  else if ([@"isIABVendorAccepted" isEqualToString:call.method])
  {
      int ID = [[call.arguments objectForKey: @"id"] intValue];
      result([NSNumber numberWithBool: [TCConsentAPI isIABVendorAccepted: ID]]);
  }
  else if ([@"isIABPurposeAccepted" isEqualToString:call.method])
  {
      int ID = [[call.arguments objectForKey: @"id"] intValue];
      result([NSNumber numberWithBool: [TCConsentAPI isIABPurposeAccepted: ID]]);
  }
  else if ([@"isIABSpecialFeatureAccepted" isEqualToString:call.method])
  {
      int ID = [[call.arguments objectForKey: @"id"] intValue];
      result([NSNumber numberWithBool: [TCConsentAPI isIABSpecialFeatureAccepted: ID]]);
  }
  else if ([@"getAcceptedCategories" isEqualToString:call.method])
  {
      result([TCConsentAPI getAcceptedCategories]);
  }
  else if ([@"shouldDisplayPrivacyCenter" isEqualToString:call.method])
  {
      result([NSNumber numberWithBool: [TCConsentAPI shouldDisplayPrivacyCenter]]);
  }
  else
  {
    result(FlutterMethodNotImplemented);
  }
}

- (NSDictionary *) getTCUserDictionary
{
    TCUser *user = [TCUser sharedInstance];
    NSMutableDictionary *dict = [[user getJsonObject] mutableCopy];
    [dict setValue: user.consentID forKey:@"consentID"];
    [dict setValue: user.consent_vendors forKey:@"consent_vendors"];
    [dict setValue: user.consent_categories forKey:@"consent_categories"];
    [dict setValue: user.external_consent forKey:@"external_consent"];
    
    return dict;
}

- (void) consentUpdated: (NSDictionary *) consent
{
    [self.channel invokeMethod: @"consentUpdated" arguments: @{@"consent" : consent, @"user" : [self getTCUserDictionary]}];
}

/**
 * Called when the user consent hasn't changed in 13 months.
 */
- (void) consentOutdated
{
    [self.channel invokeMethod: @"consentOutdated" arguments: nil];
}

/**
 * When you make a change in the JSON, there is nothing special to do.
 * But when this change is adding or removing a category, or changing an ID, we should re-display the PCM.
 */
- (void) consentCategoryChanged
{
    [self.channel invokeMethod: @"consentCategoryChanged" arguments: nil];
}

/**
 * When the JSON has significant of important changes, this callback will be called
 */
- (void) significantChangesInPrivacy
{
    [self.channel invokeMethod: @"significantChangesInPrivacy" arguments: nil];
}

- (void) firebaseConsentChanged: (NSDictionary<NSString *,NSNumber *> *) firebaseConsent
{
    id appDelegate = [UIApplication sharedApplication].delegate;
    
    if ([appDelegate respondsToSelector:@selector(firebaseConsentChanged:)])
    {
        [appDelegate performSelector:@selector(firebaseConsentChanged:) withObject: firebaseConsent];
        
        NSLog(@"firebaseConsentChanged: called on AppDelegate.");
    }
    else
    {
        NSLog(@"AppDelegate does not respond to firebaseConsentChanged:");
    }
}

@end

File: android/src/main/java/com/tagcommander/tc_consent_plugin/TCConsentPlugin.java
package com.tagcommander.tc_consent_plugin;

import android.content.Context;
import android.content.Intent;

import androidx.annotation.NonNull;

import com.tagcommander.lib.consent.ETCConsentAction;
import com.tagcommander.lib.consent.ETCConsentSource;
import com.tagcommander.lib.consent.TCConsent;
import com.tagcommander.lib.consent.TCConsentAPI;
import com.tagcommander.lib.consent.TCPrivacyCallbacks;
import com.tagcommander.lib.consent.TCPrivacyCenter;
import com.tagcommander.lib.core.TCUser;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** TCConsentPlugin */
public class TCConsentPlugin implements FlutterPlugin, MethodCallHandler, TCPrivacyCallbacks {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private Context context;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding)
  {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "tc_consent_flutter_plugin");
    channel.setMethodCallHandler(this);
    context = flutterPluginBinding.getApplicationContext();
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    @SuppressWarnings("unchecked")
    HashMap<String, Object> schemes =  new HashMap<>();

    switch (call.method)
    {
      case "setSiteIDPrivacyID":
        Integer siteID = call.argument("siteID");
        Integer privacyID = call.argument("privacyID");
        initTCConsent(siteID, privacyID, context);
        schemes.put("user", getTCUserJson());
        result.success(schemes);
        break;
      case "refuseAllConsent":
        TCConsent.getInstance().refuseAllConsent();
        result.success(null);
        break;
      case "acceptAllConsent":
        TCConsent.getInstance().acceptAllConsent();
        result.success(null);
        break;
      case "showPrivacyCenter":
        Intent PCM = new Intent(context, TCPrivacyCenter.class);
        PCM.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        context.startActivity(PCM);
        result.success(null);
        break;
      case "useAcString":
        Boolean shouldUseAcString = call.argument("useAcString");
        TCConsent.getInstance().useAcString(shouldUseAcString);
        result.success(null);
        break;
      case "initWithCustomPCM":
        Integer siteId = call.argument("siteID");
        Integer privacyId = call.argument("privacyID");
        TCConsent.getInstance().initWithCustomPCM(siteId, privacyId, context);
        schemes.put("user", getTCUserJson());
        result.success(schemes);
        break;
      case "setConsentDuration":
        Double consentDuration = call.argument("months");
        TCConsent.getInstance().setConsentDuration(consentDuration.floatValue());
        result.success(null);
        break;
      case "useCustomPublisherRestrictions":
        TCConsent.getInstance().useCustomPublisherRestrictions();
        result.success(null);
        break;
      case "saveConsentFromPopUp":
        Map<String, String> popUpConsent = call.argument("consent");
        TCConsent.getInstance().saveConsentFromPopUp(popUpConsent);
        result.success(null);
        break;
      case "saveConsent":
        Map<String, String> consent = call.argument("consent");
        TCConsent.getInstance().saveConsent(consent);
        result.success(null);
        break;
      case "saveConsentFromConsentSourceWithPrivacyAction":
        ETCConsentSource source = call.argument("source").equals("ETCConsentSource.popUp") ? ETCConsentSource.Popup : ETCConsentSource.PrivacyCenter;
        String actionValue = call.argument("action");
        ETCConsentAction action = actionValue.equals("ETCConsentAction.acceptAll") ? ETCConsentAction.AcceptAll : (actionValue.equals("ETCConsentAction.refuseAll") ? ETCConsentAction.RefuseAll : ETCConsentAction.Save);
        Map<String, String> saveConsent = call.argument("consent");
        TCConsent.getInstance().saveConsentFromConsentSourceWithPrivacyAction(saveConsent, source, action);
        result.success(null);
        break;
      case "statEnterPCToVendorScreen":
        TCConsent.getInstance().statEnterPCToVendorScreen();
        result.success(null);
        break;
      case "statShowVendorScreen":
        TCConsent.getInstance().statShowVendorScreen();
        result.success(null);
        break;
      case "statViewPrivacyPoliciesFromPrivacyCenter":
        TCConsent.getInstance().statViewPrivacyPoliciesFromPrivacyCenter();
        result.success(null);
        break;
      case "statViewPrivacyCenter":
        TCConsent.getInstance().statViewPrivacyCenter();
        result.success(null);
        break;
      case "statViewBanner":
        TCConsent.getInstance().statViewBanner();
        result.success(null);
        break;
      case "statViewPrivacyPoliciesFromBanner":
        TCConsent.getInstance().statViewPrivacyPoliciesFromBanner();
        result.success(null);
        break;
      case "getConsentAsJson":
        result.success(TCConsent.getInstance().getConsentAsJson());
        break;
      case "resetSavedConsent":
        TCConsent.getInstance().resetSavedConsent();
        schemes.put("user", getTCUserJson());
        result.success(schemes);
        break;
      case "setLanguage":
        String languageCode = call.argument("languageCode");
        TCConsent.getInstance().setLanguage(languageCode);
        result.success(null);
        break;
      case "isConsentAlreadyGiven":
        result.success(TCConsentAPI.isConsentAlreadyGiven(context));
        break;
      case "getAllAcceptedConsent":
        result.success(TCConsentAPI.getAllAcceptedConsent(context));
        break;
      case "getLastTimeConsentWasSaved":
        result.success(TCConsentAPI.getLastTimeConsentWasSaved(context));
        break;
      case "isCategoryAccepted":
        Integer category_id = call.argument("id");
        result.success(TCConsentAPI.isCategoryAccepted(category_id, context));
        break;
      case "isVendorAccepted":
        Integer vendor_id = call.argument("id");
        result.success(TCConsentAPI.isVendorAccepted(vendor_id, context));
        break;
      case "isIABPurposeAccepted":
        Integer iab_purpose_id = call.argument("id");
        result.success(TCConsentAPI.isIABPurposeAccepted(iab_purpose_id, context));
        break;
      case "isIABVendorAccepted":
        Integer iab_vendor_id = call.argument("id");
        result.success(TCConsentAPI.isIABVendorAccepted(iab_vendor_id, context));
        break;
      case "isIABSpecialFeatureAccepted":
        Integer iab_spec_feature_id = call.argument("id");
        result.success(TCConsentAPI.isIABSpecialFeatureAccepted(iab_spec_feature_id, context));
        break;
      case "getAcceptedCategories":
        result.success(TCConsentAPI.getAcceptedCategories(context));
        break;
//      case "getAcceptedGoogleVendors":
//        result.success(TCConsentAPI.getAcceptedGoogleVendors(context));
//        break;
      case "shouldDisplayPrivacyCenter":
        result.success(TCConsentAPI.shouldDisplayPrivacyCenter(context));
        break;
        case "deactivateAndroidBackButton":
          result.success(TCConsent.getInstance().deactivateBackButton = true);
          break;
      default:
        result.notImplemented();
    }
  }

  private String getTCUserJson()
  {
    try
    {
      TCUser user = TCUser.getInstance();
      JSONObject json = TCUser.getInstance().getJsonObject();
      json.put("consentID", user.consentID);
      json.put("consent_vendors", new JSONObject(user.getConsentVendors()));
      json.put("consent_categories", new JSONObject(user.getConsentCategories()));
      json.put("external_consent", new JSONObject(user.getExternalConsent()));
      return json.toString();
    } catch (JSONException e)
    {
      e.printStackTrace();
    }
    return "";
  }

  private void initTCConsent(int siteID, int privacyId, Context context)
  {
    TCConsent.getInstance().registerCallback(this);
    TCConsent.getInstance().setSiteIDPrivacyIDAppContext(siteID, privacyId, context);
  }

  @Override
  public void consentUpdated(Map<String, String> consent)
  {
    HashMap<String, Object> map = new HashMap<>();
    map.put("consent", consent);
    map.put("user", getTCUserJson());

    channel.invokeMethod("consentUpdated", map);
  }

  @Override
  public void consentOutdated() {
    channel.invokeMethod("consentOutdated", null);
  }

  @Override
  public void consentCategoryChanged()
  {
    channel.invokeMethod("consentCategoryChanged", null);
  }

  @Override
  public void significantChangesInPrivacy()
  {
    channel.invokeMethod("significantChangesInPrivacy", null);
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}
